Pascalscript x64 fixes

BranDougherty opened the pull request:
As we discussed in #60 PascalScript is not working correctly on 64 bit builds. The situation is a bit better on windows, with at least simple scripts working. On Linux, the passing of parameters from FPC to PascalScript is broken. We're able to correctly register events, but when the events are called the parameters are all wrong. I've wrapped the minor Linux fixes in {$IFDEF LINUX} because I can't test on a Mac. However, while I'm not particularly familiar with ABI stuff, if we are dealing with a System V ABI situation then those can probably be changed to {$IFDEF UNIX}.

In this PR I'd like to test various scripts and find issues across the different 64 bit platforms we currently support. A good clue for when this is ready to merge will be when we can run full-fledged ScriptCore3 scripts such as RunMode.

I'm not sure how deep this rabbit hole goes. I see various commits in the main PascalScript repo and some forks which mention 64 bit support, and we only *need* to support FPC, so this seems tractable to me.

SwamiUmami commented:
I'm trying to test these changes on a 64-bit Linux machine. I've fixed merge conflicts and dropped you a PR - https://github.com/BranDougherty/soldat/pull/1

Some observations:
- Scripts written for 1.7.1 seem to expect a `/` character in Command parameter on [Player.OnCommand](https://wiki.soldat.pl/index.php/TActivePlayer.OnCommand) and [Game.OnAdminCommand](https://wiki.soldat.pl/index.php/TGame.OnAdminCommand) events, but 1.8.0 doesn't pass the `/` prefix. I assume this is a change in 1.8.0, and it's not related to this branch, right?
- Large-scale scripts usually rely on [libdb](https://forums.soldat.pl/index.php?topic=42012.0) to store data in a database, rather than in text files. Since in 1.8.0 everything is moving towards 64-bit, maybe libdb should get a refresh as well. @XvayS would it be possible to make 64-bit versions of [libdb](https://github.com/XvayS/libdb)? This would simplify testing - people wouldn't have to install x86 dependencies to use libdb.
- I've tried writing a very simple 64-bit shared library in C++, and calls to it from server's scripts worked fine.

@BranDougherty when do you think this could stop being a draft? Do you still view this as a rabbit hole, or did you dig Soldat out of the mud, and now it's all good?

BranDougherty commented:
> I'm trying to test these changes on a 64-bit Linux machine. I've fixed merge conflicts and dropped you a PR - [BranDougherty#1](https://github.com/BranDougherty/soldat/pull/1)

Merged, cheers.

> * Scripts written for 1.7.1 seem to expect a `/` character in Command parameter on [Player.OnCommand](https://wiki.soldat.pl/index.php/TActivePlayer.OnCommand) and [Game.OnAdminCommand](https://wiki.soldat.pl/index.php/TGame.OnAdminCommand) events, but 1.8.0 doesn't pass the `/` prefix. I assume this is a change in 1.8.0, and it's not related to this branch, right?

Correct, that's not from this PR. Removing the slash feels weird and my gut reaction is to add it back but I haven't looked into it yet.
 
> * Large-scale scripts usually rely on [libdb](https://forums.soldat.pl/index.php?topic=42012.0) to store data in a database, rather than in text files. Since in 1.8.0 everything is moving towards 64-bit, maybe libdb should get a refresh as well. @XvayS would it be possible to make 64-bit versions of [libdb](https://github.com/XvayS/libdb)? This would simplify testing - people wouldn't have to install x86 dependencies to use libdb.

He hasn't made a release yet, but he sent me a 64 bit linux build which I've attached and should be good for testing purposes: [libdb_test_x64.tar.gz](https://github.com/Soldat/soldat/files/8329875/libdb_test_x64.tar.gz)

> @BranDougherty when do you think this could stop being a draft? Do you still view this as a rabbit hole, or did you dig Soldat out of the mud, and now it's all good?

There are still one or two details in the calling conventions that I'm missing, and the tests are incomplete (not 100% coverage, insufficient testing of FFI), but it works. The biggest downside is that I stuck with PascalScript (which I now view as a fundamentally flawed library), but we've made that work so far, and the backwards compatibility is nice. With just a few more fixes it will be ready for review.

BranDougherty commented:
(Note that the Windows build failures are fixed in #87, this PR has been tested on Linux and Windows.)

This PR:

* Updates to the latest RemObjects PascalScript revision.
* Fixes PascalScript 64-bit Windows and Linux FFI (#60).
* Fixes a number of ScriptCore3 bugs, including #5.
* Adds 2 major tests (see [the README](https://github.com/BranDougherty/soldat/tree/pascalscript-x64-fixes/server/scriptcore/test) for documentation):
  * BFT: An SC3 test suite designed to cover as much API surface as possible. I basically needed something like this while fixing PascalScript and patched it up to keep it around for testing purposes. It helped expose a number of bugs.
  * FFI fuzz tests: Despite my unfamiliarity with ABI-level software, I still want to have a high certainty this code works. So I wrote a [Python script](https://github.com/BranDougherty/soldat/blob/pascalscript-x64-fixes/server/scriptcore/test/gen_ffi_tests.py) to generate test cases for the custom assembly code. This code has passed hundreds of thousands of tests against FPC's own FFI on Windows and Linux.
* Fixes (and also silences) many PascalScript warnings. I believe it is genuinely not helpful to have these warnings spilled out on every soldatserver build.

This PR was *way* more work than it was worth, but now we can have a 64-bit soldatserver build and still support the tens of thousands of lines of scripts from 1.7 with minimal porting effort. Some of the benefits:

* 1.8 currently doesn't work on 32-bit builds due to GNS issues (probably not particularly difficult to fix though).
* Scripts can call out to 64-bit shared libraries.
* soldatserver scripting can run on Mac.
* We have the option to use 64-bit only closed-source libraries serverside such as [EOS](https://dev.epicgames.com/en-US/services) which is 64-bit only on Linux.

In order to test this, you should do the following on Windows and Linux:

* Run CMake with `-DADD_BFT=1` and `-DADD_FFI_FUZZ=1` set.
* Build CMake target `gen-ffi-fuzz`.
* Create a 64-bit build.
* Run soldatserver, and watch the automated portion of the tests run.
* Join the server and type `!help`, then try out the test commands listed, see that the game doesn't crash, and various SC3 functions are tested.

SwamiUmami commented:
Some FFI tests fail on my x64 Linux machine. I could be wrong, but I got the impression that passing nested records as function parameters does not work. It could also affect returning of nested records from functions. I've tried customizing the generator of fuzz tests by setting:
```python
max_members_per_struct = 2
num_structs = 10
max_args_per_function = 3
num_functions = 100
...
type_names = ['Integer', 'Int64', 'Single', 'String']
```
I'm attaching the output of `./gen_ffi_tests.py` - [BrokenFFI.zip](https://github.com/Soldat/soldat/files/8364270/BrokenFFI.zip). The failing test is test number 3:
```pascal
  a5.m0.m0 := -14;
  WriteLn(IntToStr(3));
  ret3 := fn3(a5);
  Assert(ret3.m0 = '0.28495128656800406');
```
--------------------------
I've also tried running Savage's RunMode script (thanks for libdb 0.4!). Biggest surprise for me was the lack of HWID, which is probably linked to this: https://github.com/Soldat/soldat/blob/515804371a41bcc750deade970f87a36a31f2964/client/Client.pas#L629

Apart from that, I didn't like the formatting of table when using `!elite` command. I haven't looked into it, but maybe whitespaces in chat are not displayed properly (non-monospaced font?)

I haven't spotted any critical problems. Even replay playback works.

BranDougherty commented:
> I'm attaching the output of `./gen_ffi_tests.py` - [BrokenFFI.zip](https://github.com/Soldat/soldat/files/8364270/BrokenFFI.zip). The failing test is test number 3:
> 
> ```pascal
>   a5.m0.m0 := -14;
>   WriteLn(IntToStr(3));
>   ret3 := fn3(a5);
>   Assert(ret3.m0 = '0.28495128656800406');
> ```

Hm, it's something to do with results being passed as var parameters (as documented [here](https://www.freepascal.org/docs-html/ref/refsu65.html)). FPC seems to treat all of the following prototypes as equivalent:

```pascal
type
  r0 = record
    m0: String;
  end;

  r1 = record
    m0: r0;
  end;

  r2 = record
    m0: r1;
  end;

function fn0: r2;
function fn1: r1;
function fn2: r0;
function fn3: String;
procedure p0(var a0: String);
```

I have a fix for this as a special case, but I think I'll wait a bit and try to figure it out fully before I push.

For the `TBanLists` test I just pushed a change to not check the `BannedHW/IPCount` variables. The count accessors just return the length of the relevant arrays, but after deleting a ban those arrays become sparse. I would say it was just overtesting (I wouldn't say that about the FFI fuzz tests just yet, but of course they aren't worth obsessing over either).

BranDougherty commented on server/scriptcore/ScriptCore3.pas:
```diff
@@ -397,18 +406,20 @@ procedure TScriptCore3.HandleException(E: Exception);
-        if Assigned(Self.FUnit.ScriptUnit.OnUnhandledException) and
-           Self.FUnit.ScriptUnit.OnUnhandledException(Error, E.Message, UnitName, FunctionName, Row, Col) then
-           Exit;
+        if Assigned(Self.FUnit.ScriptUnit.OnUnhandledException) then
+          if Self.CallEvent(Self.FUnit.ScriptUnit.OnUnhandledException,
+             [Error, E.Message, UnitName, FunctionName, Row, Col]) then
+            Exit;
```
You're right, FPC does have short-circuit evaluation (old Pascal didn't, some info if it matters: https://wiki.freepascal.org/$boolEval) and it seems to work left-to-right in simple situations. But try reverting to the old code here (Single condition with `and`. Make sure to change it back in the appropriate place depending on debug setting.) and trigger an unhandled exception from an SC3 script. It bypasses the check, calls `CallEvent` with null and crashes. Might be an FPC bug.

SwamiUmami replied:
I tried reverting the changes, and it does indeed end with a crash. While testing this, I've noticed another segmentation fault that occurs when handling exceptions. I've sent you a [PR for this](https://github.com/BranDougherty/soldat/pull/3).

SwamiUmami commented on server/scriptcore/ScriptCore3.pas:
```diff
@@ -631,6 +574,19 @@ function TScriptCore3.CallFunc(const Params: array of Variant;
+begin
+  try
+    Result := Self.FExec.CallEvent(Event, Params);
+  except
+    on e: TProcNotFoundException do
+      Exit;
```
When debugging a segmentation fault, I've started wondering if this exception needs to be caught separately. Why can't we just call the default `Self.HandleException`?

Same question applies to `TScriptCore3.CallFunc` above.

BranDougherty replied:
Looks like `TScriptCore3.HandleException` is primarily designed to handle exceptions thrown from PascalScript code. But `TProcNotFoundException` is an internal SC3 error. It's error message should probably be printed somewhere, though.

SwamiUmami replied:
Makes more sense now, thanks. I guess it can be ignored for now

SwamiUmami reviewed (changes requested):
Apart from the 2 problems raised in this review, should we also wait for your FFI fixes before merging this? I'm referring to [this](https://github.com/Soldat/soldat/pull/61#issuecomment-1081427948)

BranDougherty commented:
> Would you prefer lape, or a scripting engine for a completely different language (Lua, JS)?

IMHO it is a discussion for after 1.8.

> Apart from the 2 problems raised in this review, should we also wait for your FFI fixes before merging this? I'm referring to [this](https://github.com/Soldat/soldat/pull/61#issuecomment-1081427948)

Well, I did fix those issues. But at this point I am wading into FPC internals and I think it is not actually helpful to "fix" this FFI any more. Our actual exposure to FPC's internal ABI is limited to the ScriptCore3 API surface, and this code more than covers that. If anything, these fuzz tests should be reworked to create shared libraries. But IMO they have served their purpose, and I might delete them. In any case this code is more than good enough to merge.

SwamiUmami reviewed (dismissed):
I've tried the FFI fixes, and they seem to work. I haven't had a single failed test. And I agree that it doesn't make sense to keep fixing FFI, but I wouldn't delete the tests from repository.

So, I believe this is good to merge, although it would be nice if https://github.com/BranDougherty/soldat/pull/3 got merged too.

SwamiUmami reviewed (dismissed):
LGTM
